compute-budget-instruction/src/instructions_processor.rs

```diff
@@ -36,25 +37,54 @@ mod tests {
         std::num::NonZeroU32,
     };
 
-    macro_rules! test {
-        ( $instructions: expr, $expected_result: expr) => {
-            for feature_set in [FeatureSet::default(), FeatureSet::all_enabled()] {
-                test!($instructions, $expected_result, &feature_set);
-            }
-        };
-        ( $instructions: expr, $expected_result: expr, $feature_set: expr) => {
+    macro_rules! __test_inner {
+        ($instructions:expr, $feature_set:expr, |$result:ident| $body:expr) => {{
             let payer_keypair = Keypair::new();
             let tx = SanitizedTransaction::from_transaction_for_tests(Transaction::new(
                 &[&payer_keypair],
                 Message::new($instructions, Some(&payer_keypair.pubkey())),
                 Hash::default(),
             ));
 
-            let result = process_compute_budget_instructions(
+            let $result = process_compute_budget_instructions(
                 SVMMessage::program_instructions_iter(&tx),
                 $feature_set,
             );
-            assert_eq!($expected_result, result);
+
+            $body
+        }};
+    }
+
+    macro_rules! test {
+        ($instructions:expr, $expected_result:expr $(,)?) => {
+            for feature_set in [FeatureSet::default(), FeatureSet::all_enabled()] {
+                test!($instructions, $expected_result, &feature_set);
+            }
+        };
+        ($instructions:expr, $expected_result:expr, $feature_set:expr $(,)?) => {
+            __test_inner!($instructions, $feature_set, |result| {
+                assert_eq!(result, $expected_result);
+            });
+        };
+        ($instructions:expr, $expected_result:pat $(,)?) => {
+            for feature_set in [FeatureSet::default(), FeatureSet::all_enabled()] {
+                test!($instructions, $expected_result, &feature_set);
+            }
+        };
+        ($instructions:expr, $expected_result:pat if $guard:expr $(,)?) => {
+            for feature_set in [FeatureSet::default(), FeatureSet::all_enabled()] {
+                test!($instructions, $expected_result if $guard, &feature_set);
+            }
+        };
+        ($instructions:expr, $expected_result:pat, $feature_set:expr $(,)?) => {
+            __test_inner!($instructions, $feature_set, |result| {
+                assert_matches!(result, $expected_result);
+            });
+        };
+        ($instructions:expr, $expected_result:pat if $guard:expr, $feature_set:expr $(,)?) => {
+            __test_inner!($instructions, $feature_set, |result| {
+                assert_matches!(result, $expected_result if $guard);
+            });
```

Comment:
This is just me expanding these macros to support patterns.

In general, throughout these tests, you'll see me use `assert_matches` and a pattern for the program account index because:

1. It ensures that the tests will not break if the index of the program changes
2. The absolute index of the responsible program is tested in `transaction_processor.rs` so we don't need to keep testing it over and over everywhere else.